File: leetcode/C/144.二叉树的前序遍历.c
/*
 * @lc app=leetcode.cn id=144 lang=c
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

int TreeSize(struct TreeNode *root)
{
    return root == NULL ? 0 : TreeSize(root->left) + TreeSize(root->right) + 1;
}

void prevOrder(struct TreeNode *root, int *a, int *pi)
{
    if (root == NULL)
    {
        return;
    }
    a[(*pi)++] = root->val;
    //(*pi)++;

    prevOrder(root->left, a, pi);
    prevOrder(root->right, a, pi);
}
int *preorderTraversal(struct TreeNode *root, int *returnSize)
{
    int size = TreeSize(root);
    int *a = (int *)malloc(size * sizeof(int));
    int i = 0;
    prevOrder(root, a, &i);

    *returnSize = size;

    return a;
}
// @lc code=end

File: leetcode/Java/144.二叉树的前序遍历.java
/*
 * @lc app=leetcode.cn id=144 lang=java
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// 前序遍历（根左右）——递归
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();
        preorder(root,result);
        return result;
    }

    public void preorder(TreeNode root, List<Integer> result) {
        if (root == null) {
            return;
        }
        result.add(root.val); // 如果当前节点不为空，将其值添加到结果列表 result 中
        preorder(root.left, result); // 接着对当前节点的左子树进行前序遍历
        preorder(root.right, result); // 最后对当前节点的右子树进行前序遍历
    }
}

// 前序遍历（根左右），入栈顺序：根右左——迭代
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            result.add(node.val);
            if (node.right != null) {
                stack.push(node.right);
            }
            if (node.left != null) {
                stack.push(node.left);
            }
        }
        return result;
    }
}
// @lc code=end

